File: backend/app/routers/submission.py
from datetime import datetime, timezone

from fastapi import APIRouter, Depends, File, Form, HTTPException, UploadFile

from app.db import get_supabase
from app.dependencies import get_current_user
from app.models.submission import SubmissionResponse
from app.services import content_service, storage_service
from app.services import gamification_service as gam_service

router = APIRouter(tags=["submission"])

MAX_REVISIONS = 2


@router.post("/submission", response_model=SubmissionResponse)
async def create_submission(
    unit_id: str = Form(...),
    content_text: str = Form(""),
    file: UploadFile | None = File(None),
    user: dict = Depends(get_current_user),
):
    supabase = get_supabase()
    user_id = user["id"]

    if not content_service.is_checkpoint(unit_id):
        raise HTTPException(status_code=400, detail="Endpoint ini khusus unit checkpoint")

    latest = (
        supabase.table("submissions")
        .select("*")
        .eq("user_id", user_id)
        .eq("unit_id", unit_id)
        .order("submitted_at", desc=True)
        .limit(1)
        .execute()
    )
    previous = latest.data[0] if latest.data else None

    if previous and previous["status"] == "pending":
        raise HTTPException(status_code=400, detail="Submission sebelumnya masih dalam review")
    if previous and previous["status"] == "approved":
        raise HTTPException(status_code=400, detail="Unit ini sudah disetujui")
    if previous and previous["status"] == "revision_requested" and previous["revision_count"] >= MAX_REVISIONS:
        # Reflect the terminal state in the DB — otherwise this submission
        # sits at 'revision_requested' forever, indistinguishable from one
        # that's still resubmittable (PRD 4.4: "Setelah 2 revisi gagal →
        # status failed").
        supabase.table("submissions").update({"status": "failed"}).eq("id", previous["id"]).execute()
        raise HTTPException(status_code=400, detail="Maksimal revisi tercapai — ulangi unit dari awal")

    revision_count = (previous["revision_count"] + 1) if previous else 0

    content_url = None
    if file is not None:
        file_bytes = await file.read()
        content_url = storage_service.upload_submission_file(
            supabase,
            user_id,
            unit_id,
            file.filename,
            file_bytes,
            file.content_type or "application/octet-stream",
        )

    # `content_url` is stored as the raw private-bucket path (e.g.
    # "user_id/unit_id/uuid.png"), never a public URL — PRD section 6:
    # "Supabase Storage: private bucket, akses via signed URL". Callers
    # only ever see a signed, time-limited URL (see _to_response below).
    inserted = supabase.table("submissions").insert({
        "user_id": user_id,
        "unit_id": unit_id,
        "content_url": content_url,
        "content_text": content_text or None,
        "status": "pending",
        "revision_count": revision_count,
        "submitted_at": datetime.now(timezone.utc).isoformat(),
    }).execute()

    # Freeze streak decay while the reviewer is (manually) working through
    # this — PRD 3.5 / 4.2. Approval itself, and the XP/portfolio it
    # triggers, happens via a DB trigger once a human flips the status in
    # Supabase Studio (see sql/schema.sql) — there is no reviewer endpoint
    # by design (PRD section 9).
    gam_service.freeze_streak_for_checkpoint(supabase, user_id)

    return _to_response(supabase, inserted.data[0])


@router.get("/submission/my", response_model=list[SubmissionResponse])
def my_submissions(user: dict = Depends(get_current_user)):
    supabase = get_supabase()
    result = (
        supabase.table("submissions")
        .select("*")
        .eq("user_id", user["id"])
        .order("submitted_at", desc=True)
        .execute()
    )
    return [_to_response(supabase, row) for row in (result.data or [])]


def _to_response(supabase, row: dict) -> dict:
    """Swaps the stored private-bucket path for a signed URL right before
    the row leaves the API — the raw path is meaningless to a client that
    only ever holds the anon key.
    """
    if row.get("content_url"):
        row = {**row, "content_url": storage_service.get_signed_url(supabase, row["content_url"])}
    return row

File: supabase/functions/api/routes/submission.ts
// Port of backend/app/routers/submission.py.
import { Hono } from "npm:hono@4";
import { getSupabase } from "../lib/db.ts";
import { getCurrentUser, HttpError } from "../lib/auth.ts";
import { isCheckpoint } from "../lib/content.ts";
import { freezeStreakForCheckpoint } from "../lib/gamification.ts";
import { getSignedUrl, uploadSubmissionFile } from "../lib/storage.ts";

const app = new Hono();
const MAX_REVISIONS = 2;

// Swaps the stored private-bucket path for a signed URL right before the
// row leaves the API — the raw path is meaningless to a client that only
// ever holds the anon key.
// deno-lint-ignore no-explicit-any
async function toResponse(supabase: any, row: Record<string, any>) {
  if (row.content_url) {
    return { ...row, content_url: await getSignedUrl(supabase, row.content_url) };
  }
  return row;
}

app.post("/submission", async (c) => {
  const user = await getCurrentUser(c);
  const supabase = getSupabase();
  const userId = user.id as string;

  const form = await c.req.parseBody();
  const unitId = form["unit_id"] as string;
  const contentText = (form["content_text"] as string) || "";
  const file = form["file"];

  if (!isCheckpoint(unitId)) {
    throw new HttpError(400, "Endpoint ini khusus unit checkpoint");
  }

  const { data: rows } = await supabase
    .from("submissions")
    .select("*")
    .eq("user_id", userId)
    .eq("unit_id", unitId)
    .order("submitted_at", { ascending: false })
    .limit(1);
  const previous = rows?.[0] ?? null;

  if (previous?.status === "pending") throw new HttpError(400, "Submission sebelumnya masih dalam review");
  if (previous?.status === "approved") throw new HttpError(400, "Unit ini sudah disetujui");
  if (previous?.status === "revision_requested" && previous.revision_count >= MAX_REVISIONS) {
    // Reflect the terminal state in the DB — otherwise this submission sits
    // at 'revision_requested' forever, indistinguishable from one that's
    // still resubmittable (PRD 4.4: "Setelah 2 revisi gagal → status failed").
    await supabase.from("submissions").update({ status: "failed" }).eq("id", previous.id);
    throw new HttpError(400, "Maksimal revisi tercapai — ulangi unit dari awal");
  }

  const revisionCount = previous ? previous.revision_count + 1 : 0;

  let contentUrl: string | null = null;
  if (file instanceof File) {
    const bytes = await file.arrayBuffer();
    contentUrl = await uploadSubmissionFile(
      supabase,
      userId,
      unitId,
      file.name,
      bytes,
      file.type || "application/octet-stream",
    );
  }

  const { data: inserted, error } = await supabase
    .from("submissions")
    .insert({
      user_id: userId,
      unit_id: unitId,
      content_url: contentUrl,
      content_text: contentText || null,
      status: "pending",
      revision_count: revisionCount,
      submitted_at: new Date().toISOString(),
    })
    .select()
    .single();
  if (error) throw error;

  // Freeze streak decay while the reviewer is (manually) working through
  // this — approval itself, and the XP/portfolio it triggers, happens via
  // a DB trigger once a human flips the status in Supabase Studio.
  await freezeStreakForCheckpoint(supabase, userId);

  return c.json(await toResponse(supabase, inserted));
});

app.get("/submission/my", async (c) => {
  const user = await getCurrentUser(c);
  const supabase = getSupabase();
  const { data } = await supabase
    .from("submissions")
    .select("*")
    .eq("user_id", user.id as string)
    .order("submitted_at", { ascending: false });
  const results = await Promise.all((data ?? []).map((row) => toResponse(supabase, row)));
  return c.json(results);
});

export default app;
